[Codegen] Added erf ; FastMath rewrite for vector types.

keshavvinayak01 opened the pull request:
## Description
This PR introduces vectorized computations for the erf function (Picked up from [here](https://github.com/ROCm/llvm-project/blob/amd-staging/amd/device-libs/ocml/src/erfF.cl)). The scalar version has already been [merged](https://github.com/iree-org/iree/pull/20598).

Using `vector`  ops, `arith.select` and other relevant ops, we're able to implement accurate and efficient computation of erf. The logic for deciding whether to use this implementation depends on the target attribute (`isROCMBackend`), which ultimately decides whether the `FastMathPatterns` are to be used.

The plan is to extend support for other ops from the [OCML](https://github.com/ROCm/llvm-project/tree/amd-staging/amd/device-libs/ocml/src) directory.


## Testing
- Appended a new lit-test for erf in a new file `compiler/src/iree/compiler/Codegen/Common/test/math_transform_vector.mlir`, that will CHECK the expected arith/scf/math op rewrites.
- All CI quality and e2e tests are passing.

keshavvinayak01 committed:
Added Vector rewrite for erf: Fastmath; Added Lit-test for the same

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Merge branch 'iree-org:main' into main

keshavvinayak01 committed:
Merge branch 'main' of github.com:keshavvinayak01/iree

keshavvinayak01 committed:
Added more accurate erf computations

Signed-off-by: keshavvinayak01 <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Add math_transforms_vector.mlir test to build configurations

Signed-off-by: keshavvinayak01 <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
incorrect tab indent

Signed-off-by: keshavvinayak01 <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Indent again

Signed-off-by: keshavvinayak01 <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Corrected erf scalar lit for fastmath; updated coeff vals

Signed-off-by: keshavvinayak01 <keshavvinayakjha@gmail.com>

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/FastMathPatterns.cpp:
```diff
@@ -19,109 +21,115 @@ namespace {
+        createConst(0x1.1d3156p-16f), createConst(-0x1.8d129p-12f),
+        createConst(0x1.f9a6d2p-9f),  createConst(-0x1.8c3164p-6f),
+        createConst(0x1.b4e9c8p-4f),  createConst(0x1.4515fap-1f),
+        createConst(0x1.078e50p-3f)};
```
It is not clear where the coefficients come from. Maybe add a comment?

hanhanW replied:
never mind, if the above documentation is not outdated.

```
// Pattern to lower math.erf to its device lib implementation
// (from
// https://github.com/ROCm/llvm-project/blob/amd-staging/amd/device-libs/ocml/src/erfF.cl#L11)
```

keshavvinayak01 replied:
Correct, that is the source.

hanhanW replied:
Do we just use `ArrayRef` here? They are constants and they are not mutable, right?

hanhanW replied:
> @hanhanW This creates a temporary array from the initializer list, and `ArrayRef` holds a reference to that temporary array. However, the temporary array is destroyed at the end of the expression, leaving `ArrayRef` with a dangling reference. This leads to the build error:
```
error: array backing local initializer list 'coeffs1' will be destroyed at the end of the full-expression [-Werror,-Wdangling]
```
The current approach is more suitable, unless there's another immutable data structure you have in mind?

I missed that `ArrayRef` needs a temporary array, my bad. In this case, I think we can add `const` keyword which is safer.

```suggestion
    // Coefficients for |x| < 1.0.
    const SmallVector<Value, 6> coeffs1 = {
        createConst(-0x1.268bc2p-11f), createConst(0x1.420828p-8f),
        createConst(-0x1.b5937p-6f),   createConst(0x1.ce077cp-4f),
        createConst(-0x1.81266p-2f),   createConst(0x1.06eba0p-3f)};
    // Coefficients for |x| >= 1.0.
    SmallVector<Value, 7> coeffs2 = {
        createConst(0x1.1d3156p-16f), createConst(-0x1.8d129p-12f),
        createConst(0x1.f9a6d2p-9f),  createConst(-0x1.8c3164p-6f),
        createConst(0x1.b4e9c8p-4f),  createConst(0x1.4515fap-1f),
        createConst(0x1.078e50p-3f)};
```

hanhanW reviewed (commented):
Can you improve the PR description? See https://google.github.io/eng-practices/review/developer/cl-descriptions.html#informative

I haven't reviewed the details yet, and I'm waiting for the improved PR description first. 

The other question from my side is -- does it impact the accuracy? 

It looks like we have e2e tests: https://github.com/iree-org/iree/blob/33548616294b02b60467d9c7b68e494a85c7b17f/tests/e2e/math/math_ops_rocm.json#L122-L127, but we don't need to update the tolerance? It'd be good if you can have such context in PR description.

keshavvinayak01 committed:
Merge branch 'main' into keshavvinayak01/fast-math-erf

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/FastMathPatterns.cpp:
```diff
@@ -19,109 +21,116 @@ namespace {
+      acc = rewriter.create<math::FmaOp>(loc, acc, t, coeffs[i]);
+    }
+    return rewriter.create<math::FmaOp>(loc, ax, acc, ax);
+  }
```
1. Do you really need template? I don't see a requirement; template metaprogramming is usually bad, so I'd be careful when people introduce a template method. https://google.github.io/styleguide/cppguide.html#Template_metaprogramming
2.  I think we prefer `RewriterBase` in functions because functions could be refactored to other places and it can be used by other rewriters like IRRewriter, etc. In the context, people don't need to update the type to `RewriterBase`. So I'd suggest to use `RewriterBase` in the first place and it is also all you need in the function.

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/FastMathPatterns.cpp:
```diff
@@ -19,109 +21,116 @@ namespace {
+           llvm::cast<VectorType>(resultType).getElementType().isF32()))) {
+      return rewriter.notifyMatchFailure(
+          op, "Result only supports f32 or vector<f32>");
     }
```
declare `resVecType` variable. please also use it in the lambda.

The comment and error message is misleading. It also supports other vector types that have `f32` element type, but not only `vector<f32>`.

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/FastMathPatterns.cpp:
```diff
@@ -19,109 +21,116 @@ namespace {
+        SmallVector<float> values(vecType.getNumElements(), v);
+        return rewriter.create<arith::ConstantOp>(
+            loc, vecType,
+            DenseElementsAttr::get(vecType, ArrayRef<float>(values)));
```
nit: you don't need the cast. It is common to not have a cast from SmallVector to ArrayRef

keshavvinayak01 replied:
Can't make this change, leads to build error:
```
note: candidate template ignored: could not match 'ArrayRef' against 'SmallVector'
```

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/FastMathPatterns.cpp:
```diff
@@ -19,109 +21,116 @@ namespace {
-    // Replace the original op with our implementation.
-    rewriter.replaceOp(op, finalResult);
 
+    // Restore the sign
```
```suggestion
    // Restore the sign.
```

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/test/math_transform.mlir:
```diff
@@ -66,22 +66,20 @@ func.func @erf_fastmath(%arg0: f32) -> f32 attributes {
+  // CHECK-DAG: arith.constant 0.112800106 : f32
+  // CHECK-DAG: arith.constant -0.0267075151 : f32
+  // CHECK-DAG: arith.constant 0.00491381623 : f32
+  // CHECK-DAG: arith.constant -5.61801775E-4 : f32
```
Why do we keep them in order as they're consumed? The previous form looks better to me, because you can map them to the code eaiser.

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/test/math_transforms_vector.mlir:
Any reason to create a new file? I'd expect all the tests to be in `math_transform.mlir`. We split files when there are too many test cases and we know how to categorize them. I don't see any benefit for splitting the file, atm.

hanhanW reviewed (changes requested):
We recently switch to `arith::ConstantOp::create(...)` style, so please update the code as well. Thanks.

keshavvinayak01 committed:
Refactor FastErfPattern for improved polynomial evaluation and type handling

- Updated evalErfPolynomial to use specific Value types and improved FMA operation order.
- Enhanced type checks to clarify support for f32 and vector types with f32 elements.
- Removed math_transforms_vector.mlir from build configurations as it is no longer needed, add test to math_transform.mlir instead.

Signed-off-by: keshavvinayak01 <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Merge branch 'keshavvinayak01/fast-math-erf' of github.com:keshavvinayak01/iree into keshavvinayak01/fast-math-erf

keshavvinayak01 committed:
Errenous indent in math_transform.mlir

Signed-off-by: keshavvinayak01 <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Clang-formatted FastMathPatterns.cpp

Signed-off-by: keshavvinayak01 <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Removed errenous indent from math_transform.mlir

Signed-off-by: keshavvinayak01 <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Added latest lit test correction (Reorder fma)

Signed-off-by: keshavvinayak01 <keshavvinayakjha@gmail.com>

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/test/math_transform.mlir:
```diff
@@ -67,21 +68,21 @@ func.func @erf_fastmath(%arg0: f32) -> f32 attributes {
+  // CHECK-DAG: arith.constant -0.0241816975 : f32
+  // CHECK-DAG: arith.constant 0.00385781587 : f32
+  // CHECK-DAG: arith.constant -3.78677854E-4 : f32
 
   // Else region coefficients (|x| >= 1.0)
   // CHECK-DAG: arith.constant 1.69988107E-5 : f32
-  // CHECK-DAG: arith.constant -3.78677854E-4 : f32
-  // CHECK-DAG: arith.constant 0.00385781587 : f32
-  // CHECK-DAG: arith.constant -0.0241816975 : f32
-  // CHECK-DAG: arith.constant 0.106668264 : f32
-  // CHECK-DAG: arith.constant 0.634933292 : f32
-  // CHECK-DAG: arith.constant 0.128689408 : f32
+  // CHECK-DAG: arith.constant 0.128379107 : f32
+  // CHECK-DAG: arith.constant -0.376122952 : f32
+  // CHECK-DAG: arith.constant 0.112800106 : f32
+  // CHECK-DAG: arith.constant -0.0267075151 : f32
+  // CHECK-DAG: arith.constant 0.00491381623 : f32
+  // CHECK-DAG: arith.constant -5.61801775E-4 : f32
```
Is the change correct? I thought they should be in the same order?

keshavvinayak01 replied:
Yes, the old constants were incorrect too.

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/test/math_transform.mlir:
```diff
@@ -112,6 +114,56 @@ func.func @erf_fastmath(%arg0: f32) -> f32 attributes {
+  // CHECK-DAG: arith.constant dense<0.00491381623> : vector<4xf32>
+  // CHECK-DAG: arith.constant dense<-5.61801775E-4> : vector<4xf32>
+
+  // Structure checks for vector implementation
```
```suggestion
  // Structure checks for vector implementation.
```

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/FastMathPatterns.cpp:
```diff
@@ -19,109 +21,114 @@ namespace {
+      acc = rewriter.create<math::FmaOp>(loc, t, acc, coeffs[i]);
+    }
+    return rewriter.create<math::FmaOp>(loc, x, acc, x);
+  }
```
Move it outside the struct. It does not need anything from the FastErfPattern's member (and they dont exist).

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/FastMathPatterns.cpp:
```diff
@@ -19,109 +21,114 @@ namespace {
+        SmallVector<float> values(resVecType.getNumElements(), v);
+        return rewriter.create<arith::ConstantOp>(
+            loc, resVecType,
+            DenseElementsAttr::get(resVecType, ArrayRef<float>(values)));
```
I'm not sure which [constructor](https://github.com/llvm/llvm-project/blob/8ae3aeaca0b2de1702c90b120ef4b57accef9a3c/mlir/include/mlir/IR/BuiltinAttributes.h#L82) is picked in this case, but we can do it explicitly. What you're looking for is a list of `F32FloatAttr`, as the assumption is that the element type is f32. So what you want is:

```patch
-        SmallVector<float> values(resVecType.getNumElements(), v);
+        SmallVector<Attribute> values(resVecType.getNumElements(),
+                                      rewriter.getF32FloatAttr(v));
         return rewriter.create<arith::ConstantOp>(
-            loc, resVecType,
-            DenseElementsAttr::get(resVecType, ArrayRef<float>(values)));
+            loc, resVecType, DenseElementsAttr::get(resVecType, values));
```

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/FastMathPatterns.cpp:
```diff
@@ -19,109 +21,114 @@ namespace {
+    // Select between the two results using scf.if.
+    Value result;
+    if (resultType.isF32()) {
+      // For scalar types, use scf.if
```
```suggestion
      // For scalar types, use scf.if.
```

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/FastMathPatterns.cpp:
```diff
@@ -19,109 +21,114 @@ namespace {
+      // Compute t = ax * ax for the first polynomial.
+      Value t = rewriter.create<arith::MulFOp>(loc, ax, ax);
+
+      // Compute first polynomial (for |x| < 1.0)
```
```suggestion
      // Compute first polynomial (for |x| < 1.0).
```

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/FastMathPatterns.cpp:
```diff
@@ -19,109 +21,114 @@ namespace {
+      Value result1 =
+          evalErfPolynomial(ax, t, ArrayRef<Value>(coeffs1), rewriter, loc);
+
+      // Compute second polynomial (for |x| >= 1.0)
```
```suggestion
      // Compute second polynomial (for |x| >= 1.0).
```

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/FastMathPatterns.cpp:
```diff
@@ -19,109 +21,114 @@ namespace {
-      OpBuilder::InsertionGuard guard(rewriter);
+      Value t = rewriter.create<arith::MulFOp>(loc, ax, ax);
+      Value result1 =
+          evalErfPolynomial(ax, t, ArrayRef<Value>(coeffs1), rewriter, loc);
```
you dont need ArrayRef.

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/FastMathPatterns.cpp:
```diff
@@ -19,109 +21,114 @@ namespace {
+      // post-processing.
       rewriter.setInsertionPointToStart(&ifOp.getElseRegion().front());
+      Value p2 =
+          evalErfPolynomial(ax, ax, ArrayRef<Value>(coeffs2), rewriter, loc);
```
you dont need ArrayRef.

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/FastMathPatterns.cpp:
```diff
@@ -19,109 +21,114 @@ namespace {
+
+      // Compute first polynomial (for |x| < 1.0)
+      Value result1 =
+          evalErfPolynomial(ax, t, ArrayRef<Value>(coeffs1), rewriter, loc);
```
you dont need ArrayRef.

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/FastMathPatterns.cpp:
```diff
@@ -19,109 +21,114 @@ namespace {
+
+      // Compute second polynomial (for |x| >= 1.0)
+      Value p2 =
+          evalErfPolynomial(ax, ax, ArrayRef<Value>(coeffs2), rewriter, loc);
```
you dont need ArrayRef.

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/test/math_transform.mlir:
```diff
@@ -112,6 +114,56 @@ func.func @erf_fastmath(%arg0: f32) -> f32 attributes {
+
+  // CHECK-DAG: arith.constant dense<1.000000e+00> : vector<4xf32>
+
+  // Then region coefficients (|x| < 1.0)
+  // CHECK-DAG: arith.constant dense<0.128689408> : vector<4xf32>
+  // CHECK-DAG: arith.constant dense<0.634933292> : vector<4xf32>
+  // CHECK-DAG: arith.constant dense<0.106668264> : vector<4xf32>
+  // CHECK-DAG: arith.constant dense<-0.0241816975> : vector<4xf32>
+  // CHECK-DAG: arith.constant dense<0.00385781587> : vector<4xf32>
+  // CHECK-DAG: arith.constant dense<-3.78677854E-4> : vector<4xf32>
+
+  // Else region coefficients (|x| >= 1.0)
+  // CHECK-DAG: arith.constant dense<1.69988107E-5> : vector<4xf32>
+  // CHECK-DAG: arith.constant dense<0.128379107> : vector<4xf32>
+  // CHECK-DAG: arith.constant dense<-0.376122952> : vector<4xf32>
+  // CHECK-DAG: arith.constant dense<0.112800106> : vector<4xf32>
+  // CHECK-DAG: arith.constant dense<-0.0267075151> : vector<4xf32>
+  // CHECK-DAG: arith.constant dense<0.00491381623> : vector<4xf32>
+  // CHECK-DAG: arith.constant dense<-5.61801775E-4> : vector<4xf32>
```
Please update the checks if the other change is a mistake.

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/test/math_transform.mlir:
```diff
@@ -58,6 +58,7 @@ func.func @rewrite_erf(%arg0: f16) -> f16 attributes {
 
 // -----
 
+
```
delete the blank line.

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/FastMathPatterns.cpp:
```diff
@@ -19,109 +21,114 @@ namespace {
+    if (!(resultType.isF32() ||
+          (resVecType && resVecType.getElementType().isF32()))) {
+      return rewriter.notifyMatchFailure(
+          op, "Result only supports f32 or vector types with f32 element type");
```
> Also, to match error message styles commonly produced by other tools, start the first sentence with a lowercase letter, and finish the last sentence without a period, if it would end in one otherwise.

Given you're touching the code, can you help fix the style? 

```suggestion
          op, "result only supports f32 or vector types with f32 element type");
```

https://llvm.org/docs/CodingStandards.html#error-and-warning-messages

[optional] If it were me, I'd delete the above comment because we already document it in the error message.

hanhanW reviewed (commented):
LGTM, just some nits

keshavvinayak01 committed:
Addressed comments[2]

Signed-off-by: keshavvinayak01 <keshavvinayakjha@gmail.com>

hanhanW reviewed (approved):
thanks
